analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs

```diff
@@ -0,0 +1,171 @@
+                foreach (var primaryLocation in LocationIdentifiers<ClassDeclarationSyntax>(controllerSymbol))
+                {
+                    symbolEndContext.ReportIssue(Diagnostic.Create(
+                        Rule,
+                        primaryLocation,
+                        secondaryLocations.ToAdditionalLocations(),
+                        secondaryLocations.ToProperties(),
+                        responsibilityGroups.Count));
+                }
+            }
+        });
+    }
+
+    private static string BlockName(SyntaxNode block) =>
+        block switch
+        {
+            MethodDeclarationSyntax method => method.GetName(),
+            AccessorDeclarationSyntax { Parent.Parent: PropertyDeclarationSyntax property } => property.GetName(),
+            _ => null
+        };
+
+    private static IEnumerable<Location> LocationIdentifiers<T>(ISymbol symbol) where T : SyntaxNode =>
+        symbol.DeclaringSyntaxReferences.Select(x => x.GetSyntax()).OfType<T>().Select(x => x.GetIdentifier()?.GetLocation());
+
+    private static ImmutableHashSet<string> RelevantMemberNames(INamedTypeSymbol symbol)
+    {
+        var builder = ImmutableHashSet.CreateBuilder<string>();
+        foreach (var member in symbol.GetMembers())
+        {
+            switch (member)
+            {
+                // Constructors are not considered because they have to be split anyway
+                // Accessors are not considered because they are part of properties, that are considered as a whole
+                case IMethodSymbol method when !method.IsConstructor() && !method.IsStaticConstructor() && method.AssociatedSymbol is not IPropertySymbol:
+                    builder.Add(method.Name);
+                    break;
+                // Primary constructor parameters may or may not generate fields, and must be considered
+                case IMethodSymbol method when method.IsPrimaryConstructor():
+                    builder.UnionWith(method.Parameters.Where(IsService).Select(x => x.Name));
+                    break;
+                // Backing fields are excluded for auto-properties, since they are considered part of the property
+                case IFieldSymbol field when IsService(field) && !field.IsImplicitlyDeclared:
+                    builder.Add(field.Name);
+                    break;
+                case IPropertySymbol property when IsService(property):
+                    builder.Add(property.Name);
+                    break;
+            }
+        }
+
+        return builder.ToImmutable();
+    }
+
+    private static bool IsService(ISymbol symbol) =>
+        symbol.GetSymbolType() is { TypeKind: TypeKind.Interface, Name: var name } && !ExcludedWellKnownServices.Contains(name);
+
+    private static IEnumerable<SecondaryLocation> SecondaryLocations(INamedTypeSymbol controllerSymbol, List<List<string>> sets)
+    {
+        var methods = controllerSymbol.GetMembers().OfType<IMethodSymbol>().ToList();
```

Comment:
```suggestion
        var methods = controllerSymbol.GetMembers().OfType<IMethodSymbol>();
```
There's no need to evaluate this here - it's going to happen in the loop later anyways/

Reply:
By adding `ToList` it is evaluated only once, instead of once per `set` inside the `setIndex` outer-loop.

Reply:
Indeed. However, this, from my POV, is confusing the way it's written.
If Anton applies the other comment about extracting the GetMembers in a field:

```
    private static IEnumerable<SecondaryLocation> SecondaryLocations(List<List<string>> sets)
    {
        var membersLocations = MemberLocations(members.OfType<IMethodSymbol>());
        for (var setIndex = 0; setIndex < sets.Count; setIndex++)
        {
            foreach (var memberLocation in sets[setIndex].SelectMany(x => membersLocations[x]))
            {
                yield return new SecondaryLocation(memberLocation, $"Belongs to responsibility #{setIndex + 1}.");
            }
        }

        static IDictionary<string, IEnumerable<Location>> MemberLocations(IEnumerable<IMethodSymbol> members) =>
            members.GroupBy(x => x.Name).ToDictionary(x => x.Key, x => x.SelectMany(LocationIdentifiers<MethodDeclarationSyntax>));
    }
    ```
    
    Also please move the `LocationIdentifiers` method after this since it's a leaf.

Reply:
`ToList` is gone using the name-specific overload of `GetMembers`.